shared/src/main/scala/mlscript/codegen/typescript/TsTypegen.scala

```diff
@@ -0,0 +1,288 @@
+package mlscript.codegen.typescript
+
+import mlscript.utils._
+import mlscript.{ TypeVar, SourceCode, Type, Union, TypeName, Inter, Record, Tuple,
+  Top, Bot, Literal, Function, Recursive, AppliedType, Neg, Rem, Bounds, WithExtension,
+  IntLit, DecLit, StrLit, Arr }
+import mlscript.{JSBackend, JSLit}
+import mlscript.ShowCtx
+import mlscript.Typer
+import mlscript.TypeDef
+import mlscript.Terms
+import mlscript.SourceLine
+import scala.collection.mutable.{ListBuffer, Map => MutMap}
+import mlscript.codegen.Scope
+
+final case class IllFormedTsTypeError(message: String) extends Exception(message);
+
+/** Typescript typegen code builder for an mlscript typing unit
+  */
+final case class TsTypegenCodeBuilder() {
+  // store converted mlscript type definitions and terms created in mlscript
+  // use typeScope and termScope to generate non-conflicting names for both
+  private val typeScope: Scope = Scope("globalTypeScope")
+  private val termScope: Scope = Scope("globalTermScope")
+  private val typegenCode: ListBuffer[SourceCode] = ListBuffer.empty;
+
+  /** Return complete typegen code for current typing unit
+    *
+    * @return
+    *   SourceCode
+    */
+  def toSourceCode(): SourceCode = {
+    SourceCode("// start ts") +=: typegenCode
+    typegenCode += SourceCode("// end ts")
+    SourceCode.bulkConcat(typegenCode)
+  }
+
+  /** Context for converting a statement or type definition
+    *
+    * @param typeVarMapping
+    *   mapping from type variables to their clash-free parameter names
+    * @param recTypeVarSet
+    *   Set of recursive type variables for the type where context was intialized
+    * @param recTypeVarSet
+    *   Set of non-recursive type variables for the type where context was intialized
+    * @param termScope
+    *   scope for argument names
+    * @param typeScope
+    *   scope for type parameter names
+    */
+  protected case class TypegenContext(
+      val typeVarMapping: MutMap[TypeVar, SourceCode],
+      val recTypeVarSet: Set[TypeVar],
+      val nonRecTypeVarSet: Set[TypeVar],
```

Comment:
Since this will be a `lazy val` in `Type`, I don't think there's a need to store it here in the context.